numpy/core/src/multiarray/ctors.c

```diff
@@ -3519,6 +3544,26 @@ PyArray_FromBuffer(PyObject *buf, PyArray_Descr *type,
         Py_INCREF(buf);
     }
 
+#if defined(NPY_PY3K)
+    if (PyObject_GetBuffer(buf, &view, PyBUF_WRITABLE) < 0) {
+        writeable = 0;
+        PyErr_Clear();
+        if (PyObject_GetBuffer(buf, &view, PyBUF_SIMPLE) < 0) {
+            Py_DECREF(buf);
+            Py_DECREF(type);
+            return NULL;
+        }
+    }
+    data = (char *)view.buf;
+    ts = view.len;
+    /*
+     * XXX: PyObject_AsWriteBuffer does also this, but it is unsafe, as
```

Comment:
As above; here, the comment is even more confusing for those who don't think about python2 (just imaging all python2 stuff gone - this comment will probably *not* get cleaned up unless we state python2 in it!)

Reply:
This is *only* for Python 3. The `PyObject_AsWriteBuffer`/`PyObject_AsReadBuffer` functions currently available in Python 3 deal with a completely different buffer type than in Python 2, and it is only with the Python 3 `Py_buffer` that the reference to the supplying object can be released. AFAICT, the released reference is entirely useless for NumPy as ndarrays can't use it, so anything passing a buffer to an ndarray needs to know that. Maybe we should just say that we release the reference because there is nothing else to do with it. @pv Thoughts?